### ISSUE TITLE
Feature request: missing WSARecvMsg

### ISSUE THREAD

**Issue by Thomasdezeeuw** [microsoft/windows-rs#2530]
### Motivation

I'm trying to add recvmsg to socket, see https://github.com/rust-lang/socket2/pull/447. The `recvmsg(2)` equivalent on Windows seems to be [`WSARecvMsg`](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms741687(v=vs.85)), but I can't find a definition in windows-sys for it.

### Additional context

Related pr: https://github.com/rust-lang/socket2/pull/447.

**Comment by kennykerr** [comment]
Hi Thomas, this appears to be a callback function named `LPFN_WSARECVMSG`, which is available in the `windows-sys` crate.

https://learn.microsoft.com/en-us/windows/win32/api/mswsock/nc-mswsock-lpfn_wsarecvmsg

"You implement a matching WSARecvMsg callback function in your app."

**Comment by Thomasdezeeuw** [comment]
Sorry about that @kennykerr I only read (okay, skimmed) https://learn.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms741687(v=vs.85) and it doesn't mention it's a callback function signature, so I figured it was similar to `WSASendMsg`.

Do you happen to know what the `recvmsg(2)` equivalent on Windows would to be? The closest I can find is [`WSARecv`](https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsarecv), but that is missing the control data (`Control` field from [`WSAMSG`](https://learn.microsoft.com/en-us/windows/win32/api/ws2def/ns-ws2def-wsamsg)).

**Comment by kennykerr** [comment]
Not too familiar with that. @jonwis may know otherwise I'd try:

* https://stackoverflow.com/
* https://docs.microsoft.com/en-us/answers/topics/windows-api.html

**Comment by jonwis** [comment]
I've asked around. FWIW, Bing Chat suggests that `WSARecvMsg` is the equivalent of `recvmsg(2)` ... I see https://learn.microsoft.com/en-us/windows/win32/api/ws2def/ns-ws2def-wsamsg having the `Control` field. Is there sample C code using recvmsg we can look at to show how to use the WSA equivalent?

**Comment by mziller** [comment]
Yeah [WSARecvMsg](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms741687(v=vs.85)) is the API to use to get the ancillary control message data. You need to [set a socket option](https://learn.microsoft.com/en-us/windows/win32/winsock/ipproto-ip-socket-options) first to indicate what control data you want returned and then use [WSAIoctl](https://learn.microsoft.com/en-us/windows/win32/api/mswsock/nc-mswsock-lpfn_wsarecvmsg) to get the function pointer for WSARecvMsg which you can then use.

Sample is available here: https://github.com/microsoft/Windows-classic-samples/blob/main/Samples/Win7Samples/netds/winsock/recvmsg/rmmc.cpp

**Comment by jonwis** [comment]
[heart]         Jon Wiswall reacted to your message:
________________________________

**Comment by Thomasdezeeuw** [comment]
@kennykerr @jonwis @mziller thanks for the answers.

I'll make the `recvmsg` method Unix only as socket2 doesn't carry any state beyond the handle, so supporting it on Windows wouldn't be as simple as I hoped. I'll link to the example in the docs.
